File: algorithms/sjf/cpp/stl/sjf-no-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct Process {
	unsigned pid;
	unsigned at;
	unsigned bt;
	unsigned ct;
};

// ------------------------------------------------------
// SJF Algo Implementation

void findCompletionTime(std::vector<Process> &proc) {
	unsigned count = proc.size();
	unsigned fin = 0, time = 0;
	unsigned maxBT = std::max_element(
			proc.begin(),
			proc.end(),
			[] (const Process &a, const Process &b) -> bool {
			return a.bt < b.bt;
			})->bt;

	std::vector<unsigned> rem;
	std::for_each(proc.begin(), proc.end(),
			[&] (const Process p){
			rem.push_back(p.bt);
			});

	while (fin < count) {
		int idx = -1;
		unsigned min = maxBT;
		for (const auto &p: proc) {
			int pos = p.pid;
			if (rem[pos] && rem[pos] <= min && p.at <= time) {
				if (
					idx != -1 && rem[pos] == min &&
					p.at >= proc[idx].at
				   )
					continue;
				min = p.bt;
				idx = pos;
			}
		}
		if (idx == -1) {
			time++;
			continue;
		}
		time += rem[idx];
		proc[idx].ct = time;
		rem[idx] = 0;
		fin++;
	}
}

void findAvgTime(std::vector<Process> &proc) {
	findCompletionTime(proc);

	float ta{0}, w{0};

	std::cout << "PID\tAT\tBT\tCT\tTAT\tWT\n";
	for (const auto &p: proc) {
		unsigned pTA = p.ct - p.at;
		unsigned pW = pTA - p.bt;
		std::cout << p.pid << '\t' << p.at << '\t'
			<< p.bt << '\t' << p.ct << '\t'
			<< pTA << '\t' << pW << '\n';
		ta += pTA;
		w += pW;
	}

	ta /= proc.size();
	w /= proc.size();

	std::cout << std::endl;
	std::cout << "Average Waiting Time: " << w << '\n' 
		<< "Average Turn-Around Time: " << ta << '\n';
}

int main() {
	std::vector<Process> proc = {
		{0, 13, 2, 0},
		{1, 5, 1, 0},
		{2, 2, 3, 0},
		{3, 1, 1, 0},
		{4, 0, 4, 0}
	};

	findAvgTime(proc);

	std::cout << '\n';
	return 0;
}

File: algorithms/sjf/sjf-no-sort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.stream.Collectors;

class Process {
    int pid, at, bt, ct;

    public Process(int pid, int at, int bt) {
        this.pid = pid;
        this.at = at;
        this.bt = bt;
        this.ct = 0;
    }
}

class sjf {
    public static void main(String[] args) {
        ArrayList<Process> proc = new ArrayList<Process>(
                Arrays.asList(new Process(0, 1, 5), new Process(1, 0, 5), new Process(2, 0, 2), new Process(3, 2, 4)));
        float ta = 0, w = 0;

        int maxBT = Collections.max(proc, (a, b) -> (a.bt < b.bt) ? -1 : 1).bt;
        int end = 0, time = 0;
        while (end < proc.size()) {
            int idx = -1, minBT = maxBT;
            for (Process p : proc) {
                if ((p.ct == 0) && (p.bt <= minBT) && (p.at <= time)) {
                    if ((idx != -1) && (p.bt == minBT) && (p.at >= proc.get(idx).at))
                        continue;
                    minBT = p.bt;
                    idx = p.pid;
                }
            }
            if (idx == -1) {
                final int curr = time;
                idx = Collections.min(proc.stream().filter(p -> (p.at > curr)).collect(Collectors.toList()),
                        (a, b) -> (a.at < b.at) ? 1 : -1).pid;
                time = proc.get(idx).at;
            }
            time += proc.get(idx).bt;
            proc.get(idx).ct = time;
            end++;
            ta += time - proc.get(idx).at;
            w += time - (proc.get(idx).at + proc.get(idx).bt);
        }

        System.out.println("PID\tAT\tBT\tCT\tTAT\tWT");
        proc.forEach(p -> System.out.println(p.pid + "\t" + p.at + '\t' + p.bt + '\t' + p.ct + '\t' + (p.ct - p.at)
                + '\t' + (p.ct - (p.at + p.bt))));
        System.out.println("Avg. TAT: " + ta / proc.size());
        System.out.println("Avg. WT: " + w / proc.size());
    }
}